_source/_posts/2022-01-13-mobile-sso.md

```diff
@@ -0,0 +1,161 @@
+---
+layout: blog_post
+title: How to Create a Seamless Mobile SSO (Single Sign-On) Experience in iOS
+author: huan-liu
+by: internal-contributor
+communities: [mobile]
+description: "learn platform constraint on iOS platform and how to build great SSO experience around those constraints"
+tags: [ios, sso, mobile-sso, mobile]
+tweets:
+- ""
+- ""
+image: ""
+type: conversion
+---
+
+On an iPhone, when we log in to an app, we click a login button, and a website pops up to verify our credentials. Once verified, the website then redirects back to the app, and you are logged in. This familiar Single Sign-On (SSO) pattern is frequently referred to as the *redirect flow* for authentication. The use of a web browser for auth in this example was considered [a "Best Current Practice"](https://www.rfc-editor.org/rfc/rfc8252.txt) for security and usability reasons.
+
+However, in 2017, a new prompt appeared in the login flow, before you were taken to the website. The following screenshot came from an iOS app preparing to log you in with Facebook. The interface prompt informs you that a specific app, the Call of Duty game from Activision in this example, wants to use a specific website to sign you in, and it warns you about information sharing. 
+
+{% img blog/mobile-sso/ASWebAuthenticationSession-prompt-from-cod.jpeg alt:"app login prompt when authenticating with ASWebAuthenticationSession" width:"800" %}{: .center-image }
+
+There are a couple of problems with this prompt:
+
+* **Problem 1** (Where did this prompt come from?) : Many people consider this a bad user experience because the prompt looks out of place, and the wording is confusing and might alarm end users. 
+
+* **Problem 2** (Ambiguous UI message): If you implement logout functionality through the same flow, the prompt still says "Sign In", even though the user may have already clicked a "Logout" button. This too is confusing to the end user, as shown in the following screenshot. This problem is reported as [a bug on the AppAuth library](https://github.com/openid/AppAuth-iOS/issues/255), although it was designed as a security and privacy feature.
+
+
+{% img blog/mobile-sso/ASWebAuthenticationSession-on-logout.jpeg alt:"permission prompt when logging out with ASWebAuthenticationSession" width:"300" %}{: .center-image }
+
+
+In this article, I'll go into more detail on aspects of the iOS platform limitations. I'll explain why this prompt is shown and how to get around it to build a more seamless user experience. 
+
+Specifically, this article will:
+
+* explain the evolution of iOS over time, and how and why this prompt was introduced. 
+* describe various ways you can invoke a standalone or embedded browser on iOS, and explain the cookie-sharing behaviors.
+* discuss several ways to eliminate the confusing user experience issues we've identified above.
+
```

Comment:
```suggestion

**Table of Contents**{: .hide }
* Table of Contents
{:toc}

```

Optional, this is a bit low on the page, so we might just want to skip adding it.  I think adding it earlier might break up the flow of the post while explaining the problem (including the images)